src/jdk.incubator.code/share/classes/jdk/incubator/code/internal/CodeModelToAST.java

```diff
@@ -121,11 +121,15 @@ public JCTree.JCMethodDecl transformFuncOpToAST(CoreOp.FuncOp funcOp, Name metho
         }
 
         java.util.List<Value> rootValues = funcOp.traverse(new ArrayList<>(), (l, ce) -> {
-            if (ce instanceof Op op && op.result() != null && op.result().uses().size() != 1) {
-                l.add(op.result());
-            } else if (ce instanceof CoreOp.InvokeOp invokeOp && (invokeOp.invokeDescriptor().equals(M_BLOCK_BUILDER_OP)
-                   || invokeOp.invokeDescriptor().equals(M_BLOCK_BUILDER_PARAM))) {
-               l.add(invokeOp.result());
+            boolean isRoot = switch (ce) {
```

Comment:
I've rewrote the code here, as the condition `uses().size() != 1` was very subtle, and picking up things with no usages which, in principle, should not be "roots". Now the list of root values we expect is captured explicitly in the code.

Reply:
Array stores ops and return ops  were examples of ops implicitly picked up by the old code.